patroni/quorum.py

```diff
@@ -0,0 +1,316 @@
+    other state arbitrary given arbitrary changes is node availability, configuration and interrupted transitions.
+
+    To keep the invariant the rule to follow is that when increasing `numsync` or `quorum`, we need to perform the
+    increasing operation first. When decreasing either, the decreasing operation needs to be performed later.
+
+    Order of adding or removing nodes from sync and voters depends on the state of synchronous_standby_names:
+    When adding new nodes:
+        if sync (synchronous_standby_names) is empty:
+            add new nodes first to sync and then to voters when numsync_confirmed > 0
+        else:
+            add new nodes first to voters and than to sync
+    When removing nodes:
+        if sync (synchronous_standby_names) will become empty after removal:
+            first remove nodes from voters and than from sync
+        else:
+            first remove nodes from sync and than from voters. make voters empty if numsync_confirmed == 0"""
+
+    def __init__(self, leader: str, quorum: int, voters: Collection[str],
+                 numsync: int, sync: Collection[str], numsync_confirmed: int,
+                 active: Collection[str], sync_wanted: int, leader_wanted: str) -> None:
+        self.leader = leader                        # The leader according to the `/sync` key
+        self.quorum = quorum                        # The number of nodes we need to check when doing leader race
+        self.voters = CaseInsensitiveSet(voters)    # Set of nodes we need to check (both stored in the /sync key)
+        self.numsync = min(numsync, len(sync))      # The number of sync nodes in synchronous_standby_names
+        self.sync = CaseInsensitiveSet(sync)        # Set of nodes in synchronous_standby_names
+        # The number of nodes that are confirmed to reach safe LSN after adding them to `synchronous_standby_names`.
+        # We don't list them because it is known that they are always included into active.
+        self.numsync_confirmed = numsync_confirmed
+        self.active = CaseInsensitiveSet(active)    # Set of active nodes from `pg_stat_replication`
+        self.sync_wanted = sync_wanted              # The desired number of sync nodes
+        self.leader_wanted = leader_wanted          # The desired leader
+
+    def check_invariants(self) -> None:
+        """Checks invatiant of synchronous_standby_names and /sync key in DCS.
+
+        :raises `QuorumError`: in case of broken state"""
+        voters = CaseInsensitiveSet(self.voters | CaseInsensitiveSet([self.leader]))
+        sync = CaseInsensitiveSet(self.sync | CaseInsensitiveSet([self.leader_wanted]))
+
+        # We need to verify that subset of nodes that can acknowledge a commit overlaps
+        # with any subset of nodes that can achieve quorum to promote a new leader.
+        if self.voters and not (len(voters | sync) <= self.quorum + self.numsync + 1):
```

Comment:
(I'm trying to understand this code and how it complies with what is described in the docstring in regards to "keeping the invariants". Maybe it's just late for me at this point, and I'm struggling to consume it 😆 . Thoughts are welcome IAC!)

Reply:
I get that this method is supposed to determine if the nodes that are stored in the DCS sync key, hence the ones that will dictate if a failover can occur, should have an overlap with the nodes that are currently set in `synchronous_standby_names`.

I'm just having a hard time to understand how this overlap operation works and guarantees what we are looking for. I mean, I'm struggling to understand how the implemented math operation ensures that.

I was also looking to understand why we need to check if either of the sets is a subset of the other?

Under a normal situation we would have `sync` = `voters`, and the thing described in the class docstring makes sense.

Reply:
> Do we need that "+ 1"? I'm struggling to understand if that's needed and why, maybe the "<=" is enough?

Yes, we do need it:
- `numsync` is a numeric value from `synchronous_standby_names` and doesn't include the leader.
- `quorum` is a numeric value from the `/sync` key and it's meaning is - how many other nodes we need to see in order to make a decision. It's value is calculated as `len(sync) - numsync`.

Lets say out `a` is the leader name and we have `ANY 1 (b,c)` in `synchronous_standby_names`.
In the /sync key we will have: `{leader: 'a', sync: 'b,c' , quorum: 1}`

If we start performing this check we will get:
- `voters = {a,b,c}`
- `sync = {a,b,c}`
- `len(sync | voters) = 3`
- `quorum + numsync = 2`

That is, even for a normal (stable) configuration we need both, `+1` and `<=`.

When we change the topology (adding/removing/replacing nodes or changing `numsync`/`quorum` values) the rule also must be followed.

Reply:
I guess I missed the leader stuff when I was checking this operation.
I think that makes more sense now, thanks for clarifying!

Reply:
> not (nodes <= quorum)

The reason it is written like this because IMO it is easier to understand the meaning of ot.
We want to check the the value stored in the `quorum` field of the `/sync` key is high enough that we can surely identify whether the current node is one of the most up-to-date (could be count as synchronous).

If the rule is violated (`not`), that would mean the topology is in a bad shape.